submariner-operator/issue-79-kubeconfig-handling.md

```diff
@@ -0,0 +1,113 @@
+## Epic Description
+
+The various `subctl` commands support a mixture of `KUBECONFIG` environment variables, `--kubeconfig` flags, `--kubecontext` flags,
+kubeconfig files etc.
+
+We should make this consistent, ideally using `clientcmd`’s tools to handle kubeconfig.
+[Operator PR #23](https://github.com/skitt/submariner-operator/pull/23) has an old attempt at this,
+using [`BindOverrideFlags()`](https://pkg.go.dev/k8s.io/client-go@v0.23.5/tools/clientcmd#BindOverrideFlags) to provide access
+to all the features supported by `clientcmd` (see `kubectl options`).
+
+On top of that, we should align our flags with `kubectl`’s;
+beyond helping users already familiar with `kubectl` flags,
+this will make it easier to turn `subctl` into a `kubectl` plugin,
+should we choose to do so.
+This would involve turning `--kubecontext` into `--context`.
+
+## Acceptance Criteria
+
+For commands using a single context, _all_ the following should work,
+with kubeconfigs containing one or more clusters and/or contexts:
+
+```bash
+# Use the default context in the kubeconfig
+KUBECONFIG=/path/to/kubeconfig subctl foo
+subctl foo --kubeconfig /path/to/kubeconfig
+subctl --kubeconfig /path/to/kubeconfig foo
+
+# Use the specified context from the kubeconfig
+KUBECONFIG=/path/to/kubeconfig subctl foo --context bar
+subctl foo --kubeconfig /path/to/kubeconfig --context bar
+subctl --kubeconfig /path/to/kubeconfig --context bar foo
+```
+
+For commands using multiple contexts,
+there should be one “main” set of flags, with no prefix,
+and another set with a prefix describing the purpose of the context:
+
+```bash
+KUBECONFIG=/path/to/kubeconfig subctl benchmark latency --context foo --tocontext bar
+subctl benchmark latency --kubeconfig /path/to/kubeconfig --context foo --tocontext bar
+```
+
+`--kubeconfig` is preserved as-is, since users can use a single file combining all their contexts.
+However such commands should also support separate kubeconfigs, to allow using kubeconfigs which
+contain conflicting configuration (_e.g._ kubeconfigs using the same context name and user name,
+as produced by the OpenShift installer):
```

Comment:
BTW what does `kubectl` do in such case?

Does it just ignore the fact that you can use multiple conflicting configs, or does it produce an error?

Reply:
It picks the first which matches whatever context is specified, and ignores the rest, with no error. This can be used to resolve the situation; _e.g._ with multiple `admin` contexts, `kubectl config rename-context admin foo` will rename the first `admin` context, and make the next one accessible using the `admin` name.

Reply:
We can go that route then if we want to stay closely aligned, though I think it's not critical either way (ie adding another kubeconfig param)